language/move-lang/src/naming/translate.rs

```diff
@@ -387,6 +402,36 @@ fn script(context: &mut Context, escript: E::Script) -> N::Script {
     }
 }
 
+//**************************************************************************************************
+// Friends
+//**************************************************************************************************
+fn friend(context: &mut Context, mident: ModuleIdent, loc: Loc) -> Option<Loc> {
+    let current_mident = context.current_module.as_ref().unwrap();
+    if mident.0.value.address != current_mident.0.value.address {
+        // NOTE: in alignment with the bytecode verifier, this constraint is a policy decision
+        // rather than a technical requirement. The compiler, VM, and bytecode verifier DO NOT
+        // rely on the assumption that friend modules must reside within the same account address.
+        context.error(vec![
+            (loc, "Invalid friend declaration"),
+            (
+                loc,
+                "Cannot declare modules out of the current address as a friend",
+            ),
+        ]);
+        None
+    } else if &mident == current_mident {
+        context.error(vec![
+            (loc, "Invalid friend declaration"),
+            (loc, "Cannot declare the module itself as a friend"),
+        ]);
+        None
+    } else if context.resolve_module(loc, &mident) {
+        Some(loc)
+    } else {
+        None
```

Comment:
assert no errors

Reply:
I guess you mean, `assert!(context.has_errors());` if `resolve_module` failed?

Reply:
Fixed in 71ea2ce